### PR TITLE
Updating the reflect_on_all_associations method to allow for more then one symbol to be accepted.

### PR DESCRIPTION
**Pull request by mgm702**
Updating the reflect_on_all_associations method to 
allow for more then one symbol to be accepted. It
keeps current functionality as well. If method is
presented with no symbols it will return all associations.

### Motivation / Background

This Pull Request has been created because [REPLACE ME]

### Detail

This Pull Request changes [REPLACE ME]

### Checklist

Before submitting the PR make sure the following are checked:

* [ ] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [ ] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [ ] Tests are added or updated if you fix a bug or add a feature.
* [ ] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by eileencodes** [comment]
Hi @mgm702 thanks for the PR. Can you explain more your usecase for this change? When Rails uses this method internally we either want all reflections or we want specific ones. I'm not sure this change is valuable to the majority of applications and I'm having a hard time thinking of a case where I would not to be able to pass multiple types to this method.

**Comment by mgm702** [comment]
Hey @eileencodes thanks for reviewing my PR. The use case that I'm currently experiencing is at our company we are making use of the [paper_trail](https://github.com/paper-trail-gem/paper_trail) gem to restore/retain active record states and then restore them as needed. The issue with this is the associated records and restoring their state, which is why we use this method to grab the associations that don't have this issue. We developed a wrapper method that basically calls this method multiple times using the single symbol value to get the record:
```
def grab_associations(*macros)
   macros.map {|m| reflect_on_all_associations(m) }
end
```
Instead of the wrapper, which is the update I put into the reflect_on_associations method, we could make one call with the symbols provided and it would give us a quicker and more efficient result.

I also found the method confusing initially in that a plural form of a method allows for one symbol to be accepted but not for multiple. It seems to me that it should return all associations or none. Especially considering there is already a method for single associations to be returned, `_reflect_on_association(association)`, even though I recognize that's used internally. This seems to align more closely with my idea of a plural method.

**Comment by eileencodes** [comment]
`reflect_on_association` and `reflect_on_all_associations` take different arguments and don't have the same purpose. `reflect_on_association` returns reflections based on association name like `owner`. `reflect_on_all_associations` reflects on the macro like `has_many`. So it's more like "give me the reflection where the association name is owner" vs "give me reflections for all associations with the macro has many". `reflect_on_all_associations` can return multiple results even when passing `has_many` because it will give you every has many on a model.

**Comment by eileencodes** [comment]
`reflect_on_association(association)` and `reflect_on_all_associations(macro)` have different purposes and take different arguments. Think of `reflect_on_association` as "give single reflection for an association named `owner`" vs `reflect_on_all_associations` as "give me all the reflections for the `has_many` macro". The latter returns all reflections for a specific type of reflection, not a specific type of association. It can return multiple whereas `reflect_on_association` can only return one reflection.

**Comment by mgm702** [comment]
Looking at what each method returns I see what you mean by the difference between the two. It still makes it slightly confusing with the naming but I understand now why the single symbol was originally included in the `reflect_on_associations` method. I still would argue that my use case for the update is valid and it doesn't changing the existing functionality through out the Rails codebase,.

**Comment by matthewd** [comment]
Hey @mgm702,

We've had a look at this, and while the change is indeed tiny, I'd rather not introduce the overhead of the splat/array handling in a fairly internal method here.

Given there are only a handful of reflection macro types to begin with, if you need more than one, I think it's best to call `reflect_on_all_associations` with no argument, then filter it yourself. Whatever you're about to use the reflections for will involve much deeper knowledge than the presence of `macro` on the object, so I don't think there's any layering/encapsulation concern.

Basically I think I'm claiming that the optional argument is a special-case convenience for the common, but specific, case of needing just one macro-type... but it's also a trivial filter operation for the caller to apply themselves, so if requirements exceed what's supplied, that's the best way to go.

Thanks for the PR!
